doc/plugins_readme.md

```diff
@@ -0,0 +1,135 @@
+
+### Motivation
+
+Advantages of using a plugin architecture are:
+- No need to recompile entire Shogun code but only the plugin that is being developed.
+- Each plugin will have its own dependencies, so base-shogun will end up having minimal dependencies.
+- We can have 'frozen' implementation of base shogun that actually could be installed and never touched again. Plugins could change arbitrarily.
+- Scientists could use Shogun again as the building process would be much easier.
+
+### Writing a New Plugin
+
+Coming soon.
+
+### Ingredients of Plugin Architecture
+
+#### Library
+
+At the core of plugin architecture is `Library` class. It handles loading, calling and closing of plugins from shared object (`.so`) files by using [`dlopen()`](http://linux.die.net/man/3/dlopen). Loaded plugins are available as objects of `Library` class.
+
+```cpp
+// Load a plugin named "libplugin.so" as a Library class instance.
+Library library = load_library("./libplugin.so");
+
+// Load a plugin using Library constructor
+auto dup_library = Library("./libplugin.so");
+
+library == dup_library; // True
+```
+
+#### MetaClass
+
+`MetaClass` is a template class (with `typename T`) which provides an API to return a shared-pointer-like object (`Some<T>`) of `T`. `Any` object of `std::function<Some<T>()` is passed as an argument to `MetaClass` constructor as shown below:
+
+```cpp
+// erase_type() returns an Any object
+MetaClass<CSGObject> meta_class(erase_type(
+    std::function<Some<CSGObject>()>(
+        []() -> Some<CSGObject>
+        {
+            return Some<CSGObject>(new MockBaseClass);
+        }
+    )));
+```
+
+In context of a plugin, `MetaClass` objects are stored in [`Manifest`](#manifest) and are the only way for users to make objects of classes available in the plugin. See the [example](#example) below to make this more concrete.
+
+#### Manifest
+
+`Manifest` stores meta-data of `Library`. Each `Manifest` object has a description and a set meta-classes (`MetaClass`) which are responsible for creating instances of exported classes. The set of meta-classes are passed in the `Manifest` constructor in form of `std::initializer_list<std::pair<std::string,Any>>`.
+
+It is mandatory for a plugin to declare a function called `shogunManifest()` which returns an instance of `Manifest`. We provide three macros to make this easier:
+- `BEGIN_MANIFEST`: Starts manifest declaration with its description. Always immediately followed by `EXPORT`.
+- `EXPORT`: Declares class to be exported. Always use this macro between `BEGIN_MANIFEST` and `END_MANIFEST`.
+- `END_MANIFEST`: Ends manifest declaration.
+
+```cpp
+// Example of a plugin with description "Mock library".
+// The plugin exports two classes - MockClass and AnotherMockClass.
+// MockBaseClass is the parent class for the two exported classes.
+// "mock_class" and "another_mock_class" are identifiers which are used as keys
+// for the std::unordered_map used by this Manifest object under the hood.
```

Comment:
Plugin architecture is still under development but I think we only plan to support C++11 and later versions. @lisitsyn ?